doc/source/architecture.rst

```diff
@@ -0,0 +1,225 @@
+
+More relevant to this discussion is how command patterns can be used to
+implement automation APIs. If every action is a command, then that Command can
+serve as the API. So long as the Commands can be encoded in text, a scripting
+language can be used to execute them. In fact, this approach is roughly how the
+Ansys Electronics Desktop, Ansys Mechanical APDL, and the Ansys Workbench apps
+implement automation APIs for scripting.
+
+Using the command pattern as a scripting API has disadvantages. Most
+importantly, command APIs are not symmetric. The four fundamental operations on
+data within a software app are ``Create``, ``Read``, ``Update``, and
+``Delete``. These are often abbreviated as CRUD. While you can conceptually
+update, create, and delete using commands, you can not read using a command.
+This makes it difficult to "visit" the app's data model.
+
+Mechanical API implementation
+-----------------------------
+
+Mechanical's API serves the needs of both *automation* and *customization*. For
+customization, it is necessary to read the data model. For example, when
+integrating a third-party solver, you must access boundary conditions,
+geometry, material properties, mesh, and connections to properly input them to
+the solver. As discussed earlier, the command pattern does not allow this kind
+of access. Due to this fact, the Mechanical API exposes its data model directly
+to the user. This is how a hypothetical command-based API would look for
+renaming an object.
+
+.. code:: python
+
+    RenameCommand(id=100, name="New name")
+
+Instead, an API based on a data model, like mechanical's API, looks like this:
+
+.. code:: python
+
+    obj = GetObject(id=100)
+    print(obj.Name)
+    obj.Name = "New name"
+
+Notice that you could print the name by Reading a property of the object. A
+command-based API can not provide the same experience.
+
+
+PyMechanical remote interface
+-----------------------------
+
+One way to interact with the Mechanical API from Python is as a remote session.
+You can run Mechanical as a server and send a *Remote Procedure Call* (RPC).
+The server handles the RPC and returns the response to the client. Currently,
+sending commands as a string and getting the result as a string is the only
+option.
+
+PyMechanical embedded instance
+--------------------------------
+
+The embedded instance used by PyMechanical embeds an entire instance of the
+Mechanical app in-memory inside of a Python program. There is no additional
+running process associated with it. Mechanical's data model is directly
+available within Python, which means that the fully CRUD data model of the
+mechanical API can be used.
+
+
+
+Distributed systems
+===================
+
+This section contains a very basic explanation of *distributed systems*. It is
+not meant be exhaustive and rigorous, but it instead introduces just the topics
+necessary as simply as possible to help you understand the choices made by the
+designers of PyMechanical.
+
```

Comment:
This section contains a very basic explanation of *distributed systems*. It is not meant to be an exhaustive explanation, but rather an introduction to only the topics necessary to help you understand the choices made by the designers of PyMechanical.